lib/core/ipc.js

```diff
@@ -0,0 +1,90 @@
+let fs = require('./fs.js');
+let ipc = require('node-ipc');
+
+class IPC {
+
+  constructor(options) {
+    this.logger = options.logger;
+    this.socketPath = options.socketPath || fs.dappPath(".embark/embark.ipc");
+    this.ipcRole = options.ipcRole || "server";
+    ipc.config.silent = true;
+    this.connected = false;
+  }
+
+  connect(done) {
+    const self = this;
+    function connecting(_socket) {
+      let connectedBefore = false, alreadyDisconnected = false;
+      ipc.of['embark'].on('connect',function() {
+        connectedBefore = true;
+        if (!alreadyDisconnected) {
+          self.connected = true;
+          done();
+        }
+      });
+      ipc.of['embark'].on('disconnect',function() {
+        self.connected = false;
+        ipc.disconnect('embark');
+
+        // we only want to trigger the error callback the first time
+        if (!connectedBefore && !alreadyDisconnected) {
+          alreadyDisconnected = true;
+          done(new Error("no connection found"));
+        }
+      });
+    }
+
+    ipc.connectTo('embark', this.socketPath, connecting);
+  }
+
+  serve() {
+    ipc.serve(this.socketPath, () => {});
+    ipc.server.start();
+
+    this.logger.info(`pid ${process.pid} listening on ${this.socketPath}`);
+  }
+
+  on(action, done) {
```

Comment:
Each time we call `ipc.on` it creates a new listener. Could we instead do like I did in ProcessLauncher and set up the ` ipc.server.on` once and then check inside if we have subscription?

Reply:
I like this idea.

Reply:
Checked with Iuri, and since it's only used for one `on`, it's fine. However, if we end up using this class for something else and we need to subscribe to multiple events, doing what I said would improve performance.